libs/tiled-bitmap/src/cmyklayeroperations.cc

```diff
@@ -0,0 +1,331 @@
+        int sum_y = 0;
+        int sum_k = 0;
+        for (const byte* row = base; row < end; row += sourceStride)
+        {
+          for (size_t current = 0; current < 8 * 2; current += 2)
+          {
+            sum_c += row[current    ] >> 4;
+            sum_m += row[current    ] & 15;
+            sum_y += row[current + 1] >> 4;
+            sum_k += row[current + 1] & 15;
+          }
+        }
+
+        targetPtr[0] = static_cast<byte>( sum_c == 15 * 64 ? 0xF0 : (sum_c / 60) << 4)
+                     | static_cast<byte>((sum_m == 15 * 64 ? 0x0F : sum_m / 60));
+        targetPtr[1] = static_cast<byte>( sum_y == 15 * 64 ? 0xF0 : (sum_y / 60) << 4)
+                     | static_cast<byte>((sum_k == 15 * 64 ? 0x0F : sum_k / 60));
+
+        targetPtr += 2;
+      }
+
+      targetBase += targetStride;
+      sourceBase += sourceStride * 8;
+    }
+  }
+  else if (this->bps == 2)
+  {
+    for (int y = 0; y < source->height / 8; y++)
+    {
+      for (int x = 0; x < source->width / 8; x++)
+      {
+        // We want to store the average colour of the 8*8 pixel image
+        // with (x, y) as its top-left corner into targetPtr.
+        const byte* base = sourceBase + 2 * 4 * x; // start of the row
+        const byte* end = base + 8 * sourceStride; // end of the row
+
+        int sum_c = 0;
+        int sum_m = 0;
+        int sum_y = 0;
+        int sum_k = 0;
+        for (const byte* row = base; row < end; row += sourceStride)
+        {
+          for (size_t current = 0; current < 8; current++)
+          {
+            sum_c +=  row[current]         >> 6;
+            sum_m += (row[current] & 0x30) >> 4;
+            sum_y += (row[current] & 0x0C) >> 2;
+            sum_k +=  row[current] & 0x03;
+          }
+        }
+
+        targetBase[x] = static_cast<byte>(((sum_c == 192 ? 191 : sum_c) / 48) << 6)
+                     | static_cast<byte>(((sum_m == 192 ? 191 : sum_m) / 48) << 4)
+                     | static_cast<byte>(((sum_y == 192 ? 191 : sum_y) / 48) << 2)
+                     | static_cast<byte>(((sum_k == 192 ? 191 : sum_k) / 48)     );
+      }
+
+      targetBase += targetStride;
+      sourceBase += sourceStride * 8;
+    }
+  }
+  else if (this->bps == 1)
+  {
+    for (int y = 0; y < source->height / 8; y++)
+    {
+      for (int x = 0; x < source->width / 8; x++)
+      {
+        // We want to store the average colour of the 8*8 pixel image
+        // with (x, y) as its top-left corner into targetPtr.
+        const byte* base = sourceBase + 4 * x; // start of the row
+        const byte* end = base + 8 * sourceStride; // end of the row
+
+        int sum_c = 0;
+        int sum_m = 0;
+        int sum_y = 0;
+        int sum_k = 0;
+        for (const byte* row = base; row < end; row += sourceStride)
+        {
+          for (size_t current = 0; current < 8; current++)
+          {
+            if ((current & 1) == 0) {
+              sum_c +=  row[current/2]         >> 7;
```

Comment:
Yup... You definitely need a PixelIterator

Reply:
See the third bullet point in the answer to your second comment on this file.